// smm/models/ServicePoint.js
/* 
 * ServicePoint Schema
 * creteated by: Harm Meijer, harmmeiier@gmail.com
 */

module.exports = function (paramMongoose){
  
var servicePointSchema = paramMongoose.Schema({
     //_id: String,
  ct: { //creator 
    cuID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//id of the user who created the servicePoint
    oID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//organisation id
    cd: {type: Date, default: Date.now}//creation date
  },
  md: {//modified
    uID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//id of user who changed the servicePoint
    oID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//organisation id
    lu: {type:Date,required:true}//last updated date
  },
  servicePointCode: String,
  servicePointName:String,
  servicePointStatus:String,
  servicePointAddress:String,
  servicePointType:String,
  operatingHours:String,
  contactPerson:String,
  contactPhone:String,
  servicePointDescription: String,
  
});

//Model Declaration ===============================

//public Methods ===============================
servicePointSchema.statics.servicePointsByCreator = function(paramUserAccountID, paramCallback){
   return this.find({"ct.cuID" : paramUserAccountID}).sort('-ct.cd').exec(paramCallback); //static method
};


return servicePointSchema;

};
