// smm/models/Service.js
/* 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

module.exports = function (paramMongoose){
  
var serviceSchema = paramMongoose.Schema({
     //_id: String,
  ct: { //creator 
    cuID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//id of the user who created the service
    oID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//organisation id
    cd: {type: Date, default: Date.now}//creation date
  },
  md: {//modified
    uID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//id of user who changed the service
    oID: {type:paramMongoose.Schema.Types.ObjectId,required:true},//organisation id
    lu: {type:Date,required:true}//last updated date
  },
  servicePoints: [paramMongoose.Schema.Types.ObjectId],
  serviceCode: String,//this is set when service is approved
  serviceName:String,
  serviceType:paramMongoose.Schema.Types.ObjectId,//comes from servicetypes
  serviceProvider:paramMongoose.Schema.Types.ObjectId,//the company that provides the service
  price:Number,
  briefOverview:String
});

//Model Declaration ===============================

//public Methods ===============================
serviceSchema.statics.servicesByCreator = function(paramUserAccountID, paramCallback){
    return this.find({"ct.cuID" : paramUserAccountID}).sort('-ct.cd').exec(paramCallback); //static method
   
   //db.services.ensureIndex({'ct.cuID':1,'md.lu':-1})
};


return serviceSchema;

};
